### DISCUSSION TITLE
Analyze method execution time

### DISCUSSION THREAD

**Discussion by kyarovoy-softserveinc** [async-profiler/async-profiler#543] [Q&A] [answered]
Is there a way to use async-profiler to directly or indirectly (with some degree of accuracy) understand the time spent in a particular method / all methods of a profiled application (e.g. endTimeStamp - startTimeStamp)? If I'm using JFR as an output, I can theoretically analyze recordings sequentially and track when a particular method appears in a stack trace for a thread and when it disappears, which should give me endTime and startTime with 10ms accuracy (default async-profiler profiling frequency). @apangin do you think it's the right way to do it or are there any alternatives?

I'm actively using async-profiler and enjoy its low-performance impact on the profiled application compared to JProfiler. Being able to understand the execution time of the methods can be an extremely useful feature.

**Comment by toktarev** [comment]
Async profiler is implemented using the Sampling approach.
So it just shows a flame graph that describes the distribution of sampling per some events (CPU, lock, Threads).

To analyze the behavior of the particular method you should use Wall clock profiling mode.

It will show the per-thread distribution of samples and you will see what happened in the particular thread.

It will not show you the exact execution time (unline honest profiler) but you can approximately calculate it using total execution time and the percentage value for the particular method.

**Comment by apangin** [comment, marked as answer]
I suppose the question is already answered [here](https://github.com/jvm-profiling-tools/async-profiler/discussions/520#discussioncomment-1797764).
I'd only add that if you want to estimate the total time of the method (including waiting, sleeping, etc.), you need to run profiler in the wall clock mode: `-e wall`; if you want to estimate the method's on-CPU running time - use `-e cpu`.
